File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/libs/decodeLibs/Intake.java
package org.firstinspires.ftc.teamcode.libs.decodeLibs;

import com.qualcomm.robotcore.eventloop.opmode.OpMode;
import com.qualcomm.robotcore.hardware.ColorRangeSensor;
import com.qualcomm.robotcore.hardware.DcMotor;

import org.firstinspires.ftc.teamcode.constants.PowerConstants;

/** Controller for intake located at the front of the robot. */
public class Intake {
	private final OpMode bot;
	private final DcMotor leftIntakeMotor;
	private final DcMotor rightIntakeMotor;

	public Intake(OpMode opMode) {
		// Attach variable to motor hardware and set up
		bot = opMode;

		leftIntakeMotor = bot.hardwareMap.dcMotor.get("leftIntakeMotor");
		leftIntakeMotor.setMode(DcMotor.RunMode.STOP_AND_RESET_ENCODER);
		leftIntakeMotor.setMode(DcMotor.RunMode.RUN_WITHOUT_ENCODER);
		leftIntakeMotor.setDirection(DcMotor.Direction.FORWARD); // TODO: Change to REVERSE if motor rotates in wrong direction

		rightIntakeMotor = bot.hardwareMap.dcMotor.get("rightIntakeMotor");
		rightIntakeMotor.setMode(DcMotor.RunMode.STOP_AND_RESET_ENCODER);
		rightIntakeMotor.setMode(DcMotor.RunMode.RUN_WITHOUT_ENCODER);
		rightIntakeMotor.setDirection(DcMotor.Direction.REVERSE); // TODO: Change to REVERSE if motor rotates in wrong direction

	}

	public void intakeIn() {
		leftIntakeMotor.setPower(PowerConstants.INTAKE_IN_POWER);
		rightIntakeMotor.setPower(PowerConstants.INTAKE_IN_POWER);
	}

	public void intakeOut() {
		leftIntakeMotor.setPower(PowerConstants.INTAKE_OUT_POWER);
		rightIntakeMotor.setPower(PowerConstants.INTAKE_OUT_POWER);
	}

	public void intakeStop() {
		leftIntakeMotor.setPower(0.0);
		rightIntakeMotor.setPower(0.0);
	}

	/**
	 * Set a custom intake power and direction.
	 * @param direction True is forward, towards the inside of the robot; False is backward, away from the robot
	 */
	public void intakeCustom(double power, boolean direction){
		power = Math.abs(power);
		if(!direction) power = -power;
		leftIntakeMotor.setPower(power);
		rightIntakeMotor.setPower(power);
	}

	public void getTelemetryData() {
		bot.telemetry.addData("Intake state: ",
			leftIntakeMotor.getPower() > 0 && rightIntakeMotor.getPower() > 0 ? "IN" : "OUT"
		);
	}
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/season/Intake.kt
package org.firstinspires.ftc.teamcode.season;

import com.bylazar.configurables.annotations.Configurable
import com.qualcomm.robotcore.eventloop.opmode.OpMode
import com.qualcomm.robotcore.hardware.DcMotor
import com.qualcomm.robotcore.hardware.DcMotor.*
import com.qualcomm.robotcore.hardware.DcMotorSimple.*
import org.firstinspires.ftc.teamcode.utils.components.Subsystem
import kotlin.math.abs

@Configurable
/** Controller for intake located at the front of the robot.  */
class Intake(private val opMode: OpMode) : Subsystem {
  companion object {
    private var IN_POWER = 1.0;
    private var OUT_POWER = -1.0;
  }

  private val intakeMotor: DcMotor = opMode.hardwareMap.dcMotor.get("intakeMotor");
  private val gamepad = opMode.gamepad2;

  init {
    intakeMotor.mode = RunMode.STOP_AND_RESET_ENCODER;
    intakeMotor.mode = RunMode.RUN_WITHOUT_ENCODER;
    intakeMotor.direction = Direction.REVERSE;
  }

  override fun update() {
    // Switch intake direction or turn it off
    if(gamepad.right_bumper) intakeIn();
    else if(gamepad.left_bumper) intakeOut();
    else intakeStop();
  }

  fun intakeIn() {
    intakeMotor.power = IN_POWER;
  }

  fun intakeOut() {
    intakeMotor.power = OUT_POWER;
  }

  fun intakeStop() {
    intakeMotor.power = 0.0;
  }

  /**
   * Set a custom intake power and direction.
   * @param direction True is forward, towards the inside of the robot;
   * False is backward, away from the robot.
   */
  fun setIntakePower(power: Double, direction: Boolean) {
    var powerMagnitude = abs(power);

    if(!direction) powerMagnitude = -powerMagnitude;
    intakeMotor.power = powerMagnitude;
  }

  override fun getTelemetryData() {
    opMode.telemetry.addData(
      "Intake state",
      if(intakeMotor.power > 0) "IN" else "OUT"
    );
    opMode.telemetry.addData("Intake power", intakeMotor.power);
  }
}
